pandas/_libs/lib.pyx

```diff
@@ -800,8 +801,32 @@ cpdef ndarray[object] ensure_string_array(
         # short-circuit, all elements are str
         return result
 
+    from pandas.core.dtypes.common import is_float_dtype
+    if is_float_dtype(arr.dtype):  # non-optimized path
```

Comment:
Might be a smarter way of handling floats, but figured it wasn't a case worth optimizing

Reply:
do we know arr is a numpy object here?  if so just use `arr.dtype.kind == "f"`